src/wallet/wallet_tx_builder.cpp

```diff
@@ -0,0 +1,706 @@
+                        ResolvedPayment(std::nullopt, sendTo.value(), changeAmount, std::nullopt, true));
+                changeAddr = sendTo.value();
+            },
+            [&](const CScriptID& scriptId) {
+                auto sendTo = pwalletMain->GenerateChangeAddressForAccount(
+                        sendFromAccount,
+                        allowedChangeTypes({ReceiverType::P2SH}));
+                assert(sendTo.has_value());
+                resolvedPayments.AddPayment(
+                        ResolvedPayment(std::nullopt, sendTo.value(), changeAmount, std::nullopt, true));
+                changeAddr = sendTo.value();
+            },
+            [&](const libzcash::SproutPaymentAddress& addr) {
+                // for Sprout, we return change to the originating address using the tx builder.
+                changeAddr = addr;
+            },
+            [&](const libzcash::SproutViewingKey& vk) {
+                // for Sprout, we return change to the originating address using the tx builder.
+                changeAddr = vk.address();
+            },
+            [&](const libzcash::SaplingPaymentAddress& addr) {
+                // for Sapling, if using a legacy address, return change to the
+                // originating address; otherwise return it to the Sapling internal
+                // address corresponding to the UFVK.
+                std::optional<RecipientAddress> sendTo = sendFromAccount == ZCASH_LEGACY_ACCOUNT
+                    ? addr
+                    : pwalletMain->GenerateChangeAddressForAccount(
+                            sendFromAccount,
+                            allowedChangeTypes({ReceiverType::Sapling}));
+                assert(sendTo.has_value());
+                resolvedPayments.AddPayment(
+                        ResolvedPayment(std::nullopt, sendTo.value(), changeAmount, std::nullopt, true));
+                changeAddr = sendTo.value();
+            },
+            [&](const libzcash::SaplingExtendedFullViewingKey& fvk) {
+                // for Sapling, if using a legacy address, return change to the
+                // originating address; otherwise return it to the Sapling internal
+                // address corresponding to the UFVK.
+                std::optional<RecipientAddress> sendTo = sendFromAccount == ZCASH_LEGACY_ACCOUNT
+                    ? fvk.DefaultAddress()
+                    : pwalletMain->GenerateChangeAddressForAccount(
+                            sendFromAccount,
+                            allowedChangeTypes({ReceiverType::Sapling}));
+                assert(sendTo.has_value());
+                resolvedPayments.AddPayment(
+                        ResolvedPayment(std::nullopt, sendTo.value(), changeAmount, std::nullopt, true));
+                changeAddr = sendTo.value();
+            },
+            [&](const libzcash::UnifiedAddress& ua) {
+                auto zufvk = pwalletMain->GetUFVKForAddress(ua);
+                if (zufvk.has_value()) {
```

Comment:
This is synonymous with `GetUFVKIdForAddress`, which is called when `selector` is constructed:
https://github.com/zcash/zcash/blob/1b7b73b7941ad4ddc73dbcc8d767a14138bf4b31/src/keystore.h#L413-L421

Currently when `GetUFVKIdForAddress` returns `std::nullopt`, we still construct a "valid" `selector`
https://github.com/zcash/zcash/blob/1b7b73b7941ad4ddc73dbcc8d767a14138bf4b31/src/wallet/wallet.cpp#L1915-L1934

That means if we end up in this case, we will always fail to generate a change address, so if change is non-zero we will return `AddressResolutionError::ChangeAddressSelectionError`.

We should figure out why we construct a "valid" `selector` from a UA for which we don't have a corresponding UFVK; this feels like it should be an invalid selector, but maybe it was implemented this way for some other wallet RPC usage?

Reply:
I’ve removed the `pattern = ua` case when there’s no UFVK – all tests are passing, and I couldn’t find any case where we tried to do something reasonable in that situation. This allowed a bunch of other clean-up.